File: src/main/scala/com/vc/medium/L59.scala
package com.vc.medium

object L59 {
  /**
    * Input: 3
    * Output:
    * [
    *  [ 1, 2, 3 ],
    *  [ 8, 9, 4 ],
    *  [ 7, 6, 5 ]
    * ]
    **/
  def generateMatrix(n: Int): Array[Array[Int]] = {
    val arr = Array.ofDim[Int](n, n)

    if(n == 0) return arr
    var num = 1
    var rowStart = 0
    var rowEnd = n - 1
    var colStart = 0
    var colEnd = n - 1
    while(rowStart <= rowEnd && colStart <= colEnd) {
      var i = colStart
      while(i <= colEnd) {
        arr(rowStart)(i) = num
        num += 1
        i += 1
      }
      rowStart += 1

      i = rowStart
      while(i <= rowEnd) {
        arr(i)(colEnd) = num
        i += 1
        num += 1
      }
      colEnd -= 1

      i = colEnd
      while(i >= colStart) {
        arr(rowEnd)(i) = num
        i -= 1
        num += 1
      }
      rowEnd -= 1

      i = rowEnd
      while(i >= rowStart) {
        arr(i)(colStart) = num
        i -= 1
        num += 1
      }
      colStart += 1
    }
    arr
  }
}

File: src/main/java/com/vc/medium/L59.java
package com.vc.medium;

class L59 {
    public int[][] generateMatrix(int n) {
        int[][] res = new int[n][n];
        int index = 1, rowstart = 0, colstart = 0, rowend = n - 1, colend = n - 1;

        int i = 0;
        while(rowstart <= rowend && colstart <= colend) {

            i = colstart;
            while(i <= colend) {
                res[rowstart][i++] = index++;
            }
            rowstart++;

            i = rowstart;
            while(i <= rowend) {
                res[i++][colend] = index++;
            }
            colend--;

            i = colend;
            while(i >= colstart) {
                res[rowend][i--] = index++;
            }
            rowend--;

            i = rowend;
            while(i >= rowstart) {
                res[i--][colstart] = index++;
            }
            colstart++;
        }
        return res;
    }
}